src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -787,6 +872,68 @@ discover_dfuse_mounts(void)
 	return rc;
 }
 
+static int
+fetch_dfs_obj_handle(int fd, struct dfs_mt *dfs_mt, dfs_obj_t **obj)
+{
+	int                    cmd, rc;
+	d_iov_t                iov	 = {};
+	char                  *buff_obj = NULL;
+	struct dfuse_hsd_reply hsd_reply;
+	struct dfuse_il_reply  il_reply;
+
+	rc = ioctl(fd, DFUSE_IOCTL_IL, &il_reply);
+	if (rc != 0) {
+		rc = errno;
+		if (rc != ENOTTY)
+			DS_WARN(rc, "ioctl call on %d failed", fd);
+		D_GOTO(err, rc);
+	}
+
+	if (il_reply.fir_version != DFUSE_IOCTL_VERSION) {
+		D_WARN("ioctl version mismatch (fd=%d): expected %d got %d\n", fd,
+		       DFUSE_IOCTL_VERSION, il_reply.fir_version);
+		D_GOTO(err, rc);
+	}
+
+	rc = ioctl(fd, DFUSE_IOCTL_IL_DSIZE, &hsd_reply);
+	if (rc != 0) {
+		rc = errno;
+		/* Known not working for a dir */
+		if (rc != EISDIR)
+			DS_WARN(rc, "ioctl call on %d failed", fd);
```

Comment:
Previously I was stuck at,
https://github.com/daos-stack/daos/blob/master/src/client/dfs/obj.c#L574-L580
It failed due to NULL glob->iov_buf and zero glob->iov_buf_len.
I will dig deeper how to work around. I will let you know if I need your help. Thank you!

Reply:
I pushed testing code and test it with a simple code.

#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <dirent.h>
#include <assert.h>

int main(void)
{
        int fd;
        DIR *dirp;
        struct dirent *dp;

        fd = open("/dfs/dir00", O_RDONLY | O_DIRECTORY);
        assert(fd != -1);
        dirp = fdopendir(fd);
        assert(dirp != NULL);

        while ((dp = readdir(dirp)) != NULL) {
            printf("%s\n", dp->d_name);
        }

        closedir(dirp);
        close(fd);
        return 0;
}

I got issues in readdir().
https://github.com/daos-stack/daos/blob/master/src/client/dfs/readdir.c#L57
Error message when executing this line,
**mem  ERR  src/gurt/hash.c:1273 d_hhash_link_lookup() NULL PTR type key.**

rc = daos_obj_list_dkey(obj->oh, DAOS_TX_NONE, &number, kds, &sgl, anchor, NULL);
rc returns -1002 (Invalid handle).
Looks like the dir handle was not inserted into a hash table somewhere.

@ashleypittman Do you have any suggestion? Thank you very much!

Reply:
@ashleypittman I am checking the code of dfs_dup().
https://github.com/daos-stack/daos/blob/master/src/client/dfs/obj.c#L477-L496
Looks like daos_obj_open() is required for a dir although a regular file needs daos_array_local2global() and daos_array_global2local(). 
It sounds like returning error EISDIR is expected before calling handle_dsize_ioctl(), right?
https://github.com/daos-stack/daos/blob/master/src/client/dfuse/ops/ioctl.c#L441-L442

Reply:
Can you elaborate on that?  It's not at all clear to me why files and directories would be different here.  For `ioil` then passing the directory handle isn't of benefit so I've never had reason to test this code before.

Perhaps a standalone test case of using the ioctl to fetch the handle and call dfs_iterate() on it would be helpful here.

Reply:
@ashleypittman Probably we need to call daos_obj_open() for a dir similar to daos_array_open_with_attr() for a regular file?
https://github.com/daos-stack/daos/blob/master/src/client/dfs/obj.c#L674-L675

Reply:
https://github.com/daos-stack/daos/pull/13294/commits/ee104e25084caffe6a702dec0d8e4fa7ecb5cd6d
daos_obj_open() is called for a dir similar to daos_array_open_with_attr() for a regular file inside dfs_obj_global2local(). Now the dir handle is valid.

Reply:
I've checked your updates to `dfs_obj_local2global` locally and they seem to work so I'm not sure of the error here.

Reply:
Thank you for checking. Now the issue in retrieving dir handle from dfuse by fetch_dfs_obj_handle is resolved.